Finish pile_map support enough for a basic transformer example

axch opened the pull request:
The remaining operations to cover were
- Slicing a non-ragged axis when ragged axes are present
- Einsum where both tensor and contraction dimensions are ragged

Also improved the error message that happens when trying to broadcast a ragged axis to a rectangular one

axch committed:
Better error message when broadcasting ragged to static shape.

Co-authored-by: Matthew Johnson <mattjj@google.com>

axch committed:
Test and implement ragged slicing.

This touches _gather_batching_rule because slicing is implemented as a
gather, but we only test the case exercised by the slice that occurs
in our test transformer model, namely the unstack operation
  q, k, v = qkv
(which turns into three slices on an non-batched and non-ragged axis).

Co-authored-by: Matthew Johnson <mattjj@google.com>

axch committed:
Force RaggedAxis axes to be sorted at rest, so that == is more reliable.

axch committed:
Test a different configuration of einsum.

This version stresses my transpose_ragged_axes method, which, it
seems, was interpreting the permutation the wrong way.  Fixed.

axch committed:
More type annotations, and make transpose_ragged_axes a top-level function instead of a method.

Keep move_stacked_axis as a method because it's a type-specific
version of a top-level function of the same name that already exists.

mattjj commented on tests/dynamic_api_test.py:
```diff
@@ -1627,7 +1639,7 @@ def func(size):
     self.assertAllClose(p.data, data)
 
-  def test_ragged_einsum(self):
+  def test_einsum_ragged_tensor(self):
```
```suggestion
  def test_einsum_ragged_array(self):
```

😃

axch replied:
I meant that the tensor dimension is ragged, in both cases.

axch replied:
Clarified that in the names.

mattjj commented on tests/dynamic_api_test.py:
```diff
@@ -1640,6 +1652,31 @@ def fprop_layer(x_size):
     self.assertRegex(str(p.aval), r'Var[0-9]+:3 => i32\[3,bint\{≤5\}\[3\] with value: \[3 1 4\]\.Var[0-9]+,2,7\]')
     self.assertEqual(p.data.shape, (3, 3, 5, 2, 7))
 
+  def test_einsum_ragged_tensor_and_contract(self):
```
```suggestion
  def test_einsum_ragged_array_and_contract(self):
```

😃

mattjj commented on jax/_src/lax/slicing.py:
```diff
@@ -1287,10 +1295,17 @@ def _gather_batching_rule(batched_args, batch_dims, *, dimension_numbers,
+    # `RaggedAxis` as long as we are not gathering from any ragged axis, and as
+    # long as we have no collapsed_slice_dims.  (The latter could require
+    # adjusting the `ragged_axes` field of `operand_bdim`).  Should we put a
+    # check to confirm?
```
Let's definitely put a check! Any reason not to? (We can raise NotImplemented in unhandled cases iiuc).

axch replied:
Shape prediction improved and check added.

mattjj reviewed (approved):
Nice!

axch committed:
Abstract out and reuse the gather_shape_computation to predict which axes will end up ragged.

This should resolve worries about silently wrong metadata about
pile_mapped gather, but gather is complicated so it's hard to be sure.

axch committed:
Clearer test names.

yashk2810 commented:
Side note: maybe rename pile_map to something else? Just google `pile` and you'll see what I am talking about ;)

axch commented:
@yashk2810 Yes, that's on the agenda.

mattjj reviewed (approved):
Woo!
